core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -36,43 +37,55 @@
     public final ParseGraph order;
     public final BigInteger offset;
     public final Source source;
+    public final ImmutableList<BigInteger> iterations;
```

Comment:
Something that now could happen is that `closeBranch(Token)` is called with a `Token` that was **not** the one used for `addBranch(Token)`. This would of course be a bug, but since these methods are `public`, we might need to guard for it. We could fix that by keeping track which token is used for the iteration. E.g. something like:

```java
public final ImmutableList<Pair<Token, BigInteger>> iterations;
```

Whenever we call `closeBranch(Token)`,  we can verify that the `Token` of the current iteration is indeed the one being passed. We could even pass `Token` to `iter()` and be sure that we are iterating the expected `Token` (but that might be overkill).

(Note: We have a `Pair` class in `Selection`, but it is not generic.)

Reply:
Good suggestion! I'm usually not a fan of using a Pair class, but I cannot think of an alternative, other then creating a new class, which would most likely be implemented as the Pair class. ;)

Reply:
Should it in this case be an identity comparison instead of a (semantic) equality?

Reply:
Oh that's a good point... If we want to really make sure we are using the _same_ object, instead of just an equal one.